### PR TITLE
Added validation to addresses tab, section: "DHCP Entries".

### PR DESCRIPTION
**Pull request by ghost**
Validations:
- picked ips must be from detected addresses,
- picked MACs must be from detected components (ethernets),
- picked MACs must be not assigned to other device in DHCP entry.

### CODE REVIEW

File: src/ralph/ui/forms/addresses.py [hunk 1]
```diff
@@ -293,6 +293,91 @@ def clean(self):
...
 
 
+class DHCPFormSet(DHCPBaseFormSet):
+    def __init__(self, *args, **kwargs):
+        self.device = args[3]
+        return super(DHCPFormSet, self).__init__(*args, **kwargs)
+
+    def check_ip_is_from_detected(self, form, form_idx, valid_addresses):
+        valid = True
+        if form['ip'].value() not in valid_addresses:
```
Comment (mkurek) [inline src/ralph/ui/forms/addresses.py:303]: does this work with empty fields?

Comment (ghost) [inline src/ralph/ui/forms/addresses.py:303]: yup, it works.

**Comment by mkurek** [comment]
:ok:

File: src/ralph/ui/forms/addresses.py [hunk 2]
```diff
@@ -293,6 +293,91 @@ def clean(self):
...
+            ip_errors = self.errors[form_idx].setdefault('ip', [])
+            error_msg = _(
+                'Ip address should be from detected addresses'
```
Comment (andrzej-jankowski) [inline src/ralph/ui/forms/addresses.py]: `'Ip address should be from detected addresses: '`

File: src/ralph/ui/forms/addresses.py [hunk 3]
```diff
@@ -293,6 +293,91 @@ def clean(self):
...
+            ip_errors.append(error_msg)
+            valid = False
+        return valid
+
+    def check_mac_from_device_components(self, form, form_idx, valid_macs):
+        valid = True
+        if form['mac'].value() not in valid_macs:
+            mac_errors = self.errors[form_idx].setdefault('mac', [])
+            error_msg = _(
+                'MAC address should be from detected ethernets:'
```
Comment (andrzej-jankowski) [inline src/ralph/ui/forms/addresses.py]: `'MAC address should be from detected ethernets: '`

File: src/ralph/ui/forms/addresses.py [hunk 4]
```diff
@@ -293,6 +293,91 @@ def clean(self):
...
+            form = self.forms[i]
+            if self.can_delete:
+                if self._should_delete_form(form):
+                    # This form is going to be deleted so any of its errors
+                    # should not cause the entire formset to be invalid.
+                    continue
+            if form['mac'] in invalid_mac2device.keys():
+                mac_errors = self.errors[i].setdefault('mac', [])
+                error_msg = _(
+                    'This MAC is used by other device with sn: "{}"'.format(
```
Comment (ghost) [inline src/ralph/ui/forms/addresses.py]: id or better link

File: src/ralph/ui/forms/addresses.py [hunk 5]
```diff
@@ -293,6 +293,91 @@ def clean(self):
...
+            if not valid:
+                forms_valid = False
+
+        # check if picked MACs from dhcp-entries it's not assigned to other
+        # device
+        used_dhcp_entries = DHCPEntry.objects.filter(mac__in=forms_macs).all()
+        invalid_mac2device = {
+            k: v for k, v in
+            Ethernet.objects.exclude(device=self.device).filter(
+                mac__in=used_dhcp_entries,
```
Comment (ghost) [inline src/ralph/ui/forms/addresses.py]: mac vs dhcp entry

**Comment by andrzej-jankowski** [comment]
:ok:

**Pull request merged**